### DISCUSSION TITLE
(Another) mysterious MissingAPI error

### DISCUSSION THREAD

**Discussion by dusty-phillips** [dexie/Dexie.js#1411] [General]
Just discovered something is still odd with fake-indexeddb and the latest Dexie release candidate.

It seems to be able to add data to the database and retrieve it just fine, but actually there is an error on `put` and `add` calls.

Best described with a repro:

```javascript
import Dexie from "dexie"; // 3.2.0-rc.2
import indexedDB from "fake-indexeddb"; // 3.1.3
import IDBKeyRange from "fake-indexeddb/lib/FDBKeyRange.js";

const db = new Dexie("MyDatabase", {
  indexedDB: indexedDB,
  IDBKeyRange: IDBKeyRange,
});

db.version(1).stores({
  friends: "name,shoeSize",
});

// If we don't catch the error on put...
db.friends.put({ name: "Nicolas", shoeSize: 8 });
// The data is added and we can get it
db.friends.get("Nicolas").then(console.log); // { name: 'Nicolas', shoeSize: 8}

// But if we add a catch clause to the put...
db.friends.put({ name: "Nicolas", shoeSize: 8 }).catch(console.log);
// We get MissingAPIError
```

I'm guessing that Dexie is trying to access the global indexedDB at some point, rather than the one passed into the options on the constructor. Here's the trace:

```
DexieError [MissingAPIError]: IndexedDB API missing. Please visit https://tinyurl.com/y2uuvskb
    at cmp (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4035:15)
    at addRange (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4135:18)
    at file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4127:29
    at Array.forEach (<anonymous>)
    at RangeSet.addKeys (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4127:14)
    at file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4297:44
    at file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:1159:23
    at callListener (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:881:19) {
  _e: Error
      at getErrorWithStack (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:286:12)
      at new DexieError (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:385:19)
      at cmp (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4035:15)
      at addRange (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4135:18)
      at file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4127:29
      at Array.forEach (<anonymous>)
      at RangeSet.addKeys (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4127:14)
      at file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:4297:44
      at file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:1159:23
      at callListener (file:///Users/dustyphillips/Desktop/Code/dexie-wtf/node_modules/dexie/dist/modern/dexie.mjs:881:19),
  inner: null
}
```

**Comment by dfahlander** [comment]
Ok, thanks!

**Comment by dfahlander** [comment]
This is because RangeSet class depends on indexedDB.cmp(). Could put the RangeSet constructor on the db instance and generate one per db instance (as doing with Transaction, Table, Collection etc). I am though very tempted to do a breaking change here and stop accepting the {indexedDB, IDBKeyRange} as options because static methods like Dexie.delete() would anyway need to be configured by setting the static deps in Dexie.dependencies globally.

```js
import indexedDB from "fake-indexeddb"; // 3.1.3
import IDBKeyRange from "fake-indexeddb/lib/FDBKeyRange.js";

// Configure DOM dependencies on static prop
Dexie.dependencies.indexedDB = indexedDB;
Dexie.dependencies.IDBKeyRange = IDBKeyRange;

// ...is needed before calling static method Dexie.delete():
await Dexie.delete("MyDatabase");

```

Might need to bump version to 4.0 if so. The thing is - is it worth it having a lot of plumping code just to support per-instance IDBFactory? Maybe it is? Thoughts?

**Comment by dusty-phillips** [comment]
I can only think of one use case for per-instance configuration: parallel unit tests. But I think  that can be approximated by having the one global dependency and using different database names for each dexie instance. I’ll try it and get back to you.

From a marketing/social perspective, having an excuse to launch Dexie 4.0 with dexie-cloud 1.0 might help garner some attention. 😉

**Comment by dusty-phillips** [comment]
Confirmed: copy-pasting your dependency setup code and appending a `Math.random` to the name solves the parallel test case.

**Comment by dfahlander** [comment]
Thanks for your thoughts! I'll keep this issue open for some days before taking a final decision. Anyone feel free to weigh in. To sum up:
* Dexie has since one of its earliest versions, had the option to replace the DOM dependency of indexedDB with another providers (such as indexedDbShim or fakeIndexedDB) - this way it can run in node (for example in unit tests).
* There have been two ways of configuring the DOM dependencies:

    1. By setting static properties `Dexie.dependencies.indexedDB` and `Dexie.dependencies.IDBKeyRange` to the alternate implementations
    2. By providing {indexedDB, IDBKeyRange} in options to Dexie constructor.

* The latter one is a bit more elegant but does not solve how to replace indexedDB in static methods like `Dexie.delete()`, `Dexie.exists()` and `Dexie.getDatabaseNames()`.
* This issue has been a bug since [dexie@3.1.0-alpha.1](https://github.com/dfahlander/Dexie.js/releases/tag/v3.1.0-alpha.1) (the one with liveQuery()) that has a new dependency to indexedDB.cmp() that needs to be accessed in situations when we do not have a db instance - so we don't know which implementation of indexedDB.cmp() to use.
* We could replace the dependency to indexedDB.cmp() by using a JS implemention of the same. But that would increase the code size and could potentially introduce bugs in case the JS implementation would differ from the native in some way - so we would need to have it yet another time in beta or rc before releasing it stable.

**Comment by dfahlander** [comment]
Replacing indexedDB.cmp() with a spec compliant plain javascript implementation of indexedDB.cmp() executes between 5 and 8 times faster depending on browser. This fact could speak for solving this issue by using the plain JS function instead - but it would increase the bundle size of dexie.min.js.

Chrome:

| Implementation | Number of iterations | Time taken |
|---|---|---|
| indexedDB.cmp() | 1_000_000 | 570 ms |
| plain JS version | 1_000_000 | 72 ms |

Firefox:

| Implementation | Number of iterations | Time taken |
|---|---|---|
| indexedDB.cmp() | 1_000_000 | 290 ms |
| plain JS version | 1_000_000 | 56 ms |

Safari:

| Implementation | Number of iterations | Time taken |
|---|---|---|
| indexedDB.cmp() | 1_000_000 | 108 ms |
| plain JS version | 1_000_000 | 27 ms |


Some conclusions:

* If we'd replace indexedDB.cmp() with the plain JS implementation, it would increase dexie.min.js with 886 characters and the final gzip would increase with 342 bytes - from 25.1 kb to 25.4 kb.
* This issue would be resolved if we'd do that without having to do the breaking change since RangeSet used to track live queries would not have a dependency to indexedDB anymore.
* The speed of liveQuery() could theoretically benefit from it (when comparing results of hundreds of thousands records).

So this boils down to whether it would be worth the increase of the final bundle size.

**Comment by dfahlander** [comment]
I could optimize the code further and make it faster for all type comparisions while also reducing code size a bit. Added unit tests for the new cmp function. PR is #1412.

Implemetation of cmp: [cmp.ts](https://github.com/dfahlander/Dexie.js/blob/js-cmp/src/functions/cmp.ts)

Now after some new measures when comparing string, we have an implementation that is about 30 times faster than the native and with reduced code size:

Chrome:

| Implementation | Number of iterations | Time taken (string) | Time taken (number) | Time taken (Uint8Array) |
|---|---|---|---|---|
| indexedDB.cmp() | 1_000_000 | 550 ms | 503 ms | 790 ms |
| plain JS version | 1_000_000 | 17.5 ms | 13 ms | 27 ms |
| *factor faster* | | *31* | *39* | *29* |

Firefox:

| Implementation | Number of iterations | Time taken (string) |
|---|---|---|
| indexedDB.cmp() | 1_000_000 | 370 ms |
| plain JS version | 1_000_000 | 50 ms |

Safari:

| Implementation | Number of iterations | Time taken (string) |
|---|---|---|
| indexedDB.cmp() | 1_000_000 | 110 ms |
| plain JS version | 1_000_000 | 22 ms |

Bundle sizes

| Dexie version | Minified | Minified & gzipped |
|---|---|---|
| dexie@3.0.3 | 71.13 kb | 22.1 kb |
| dexie@3.2.0-rc.2 (with liveQuery()) | 78.3 kb | 25.1 kb |
| dexie@3.2.0-... with also JS cmp | 79 kb | 25.3 kb |
